protocol_decoder/src/decoding.rs

```diff
@@ -281,7 +293,45 @@ impl ProcessedBlockTrace {
                 let dummy_txn = create_dummy_gen_input(other_data, initial_trie_state, 0);
                 gen_inputs.insert(0, dummy_txn);
             }
-            _ => (),
+            _ => dummys_added = false,
+        }
+
+        dummys_added
+    }
+
+    /// The withdrawls are always represented as a single "dummy" txn at the end
+    /// of the block. However, if no dummies have already been added, then
+    /// we need to append one to the end. If dummies have been added, then
+    /// add it to the last one.
+    fn add_withdrawls_to_txns(
+        txn_ir: &mut Vec<TxnProofGenIR>,
+        other_data: &OtherBlockData,
+        final_trie_state: &PartialTrieState,
+        withdrawals: Vec<(Address, U256)>,
+        dummies_already_added: bool,
+    ) {
+        if withdrawals.is_empty() {
+            return;
+        }
+
+        match dummies_already_added {
+            false => {
+                // Guarenteed to have a real txn.
+                let final_ir = txn_ir.last().unwrap();
+
+                // Dummy state will be the state after the final txn.
+                let withdrawl_dummy =
+                    create_dummy_gen_input(other_data, final_trie_state, final_ir.txn_idx + 1);
+
+                // If we have no actual dummy txns, then we create one and append it to the end
+                // of the block.
+                txn_ir.push(withdrawl_dummy);
+            }
+            true => {
+                // If we have dummy txns (note: `txn_ir[1]` is always a dummy txn in this case),
+                // then this dummy will get the withdrawls.
+                txn_ir[1].gen_inputs.withdrawals = withdrawals;
```

Comment:
I guess this is fine to keep as is, but we may want to emphasize that dummy txns aren't really dummy txns, but more like dummy/empty proofs, as there is no state transition involved. As such, I'd see the withdrawals proof (with no signed_txn) as different from actual padding proofs (the dummy/empty ones currently), but maybe that's just my way of seeing things.
This could probably be rediscussed for a global PR on documentation across the crate though.